packages/frontend/src/Happy/Frontend.hs

```diff
@@ -0,0 +1,57 @@
+module Happy.Frontend where
+  
+import Happy.Grammar
+import Happy.Frontend.AbsSyn
+import Happy.Frontend.Mangler
+import Happy.Frontend.PrettyGrammar
+import Happy.Frontend.Parser
+import Happy.Frontend.ParseMonad.Class
+import System.Exit
+import System.IO
+
+parseYFileContents :: String -> ParseResult AbsSyn
+parseYFileContents contents = runFromStartP ourParser contents 1
+
+mangleAbsSyn :: AbsSyn -> String -> ParseResult Grammar
+mangleAbsSyn abssyn filename = first unlines (mangler filename abssyn)
+  where first f (Left a) = Left (f a)
+        first _ (Right b) = Right b
+
+writePrettyFile :: Maybe String -> AbsSyn -> IO ()
+writePrettyFile location abssyn = do
+  let out = render (ppAbsSyn abssyn) in
+    case location of
+      Just s   -> writeFile s out >>
+        hPutStrLn stderr ("Production rules written to: " ++ s)
+      Nothing  -> return ()
```

Comment:
We don't want random IO stuff in the library. This should stay in the exe.

Reply:
True. The reason I put this (and `mangleAbsSyn` below) in `Frontend` is to reduce the number of exposed modules of `happy-frontend`, but yeah this shouldn't really be that relevant.